### PR TITLE
Add support for Molmo

### LINKED ISSUE #70: Please add support for Molmo
This is SOTA open source models based on qwen2 vl - With unprecedented accuracy it can mark the coordinates of objects on screenshots, which distinguishes it from other VLMs. You can make amazing projects of autonomous computer control, but so far there is access only to FP-32 scales that simply do not fit into the RAM of even a 36GB Mac

### PR DESCRIPTION
**Pull request by Blaizzy**
Closes #70

### CODE REVIEW

**Comment by Blaizzy** [comment]
@awni 

Does fast rope support interleave operation?

```python
def rotate_every_two(self, x: mx.array) -> mx.array:
        B, nh, T, hs = x.shape
        x = x.reshape(B, nh, T, hs // 2, 2)
        x1, x2 = x[:, :, :, :, 0], x[:, :, :, :, 1]
        x = mx.stack([-x2, x1], axis=-1)
        return x.reshape(B, nh, T, hs)
```

**Comment by awni** [comment]
Yes, that's the traditional RoPE. You can do `nn.RoPE(traditional=True)`

**Comment by Blaizzy** [comment]
When I use traditional `False:

Screenshot 2024-11-15 at 1 14 10 AM

When I use the custom Rope.
Screenshot 2024-11-15 at 1 16 22 AM

Implementation:
```python
class RotaryEmbedding(nn.Module):
    """
    MLX implementation of Rotary positional embeddings (RoPE) without caching.
    """

    def __init__(self, config: TextConfig):
        super().__init__()
        self.config = config

    def get_rotary_embedding(self, seq_len: int) -> Tuple[mx.array, mx.array]:
        # Computing rotary embeddings
        dim = self.config.d_model // self.config.n_heads
        inv_freq = 1.0 / (self.config.rope_theta ** (mx.arange(0, dim, 2) / dim))
        seq = mx.arange(seq_len, dtype=mx.float32)
        freqs = mx.einsum("i,j->ij", seq, inv_freq)

        if self.config.rope_impl == "interleave":
            # Simulate repeat_interleave
            positions = mx.concatenate(
                [freqs.reshape(-1, 1), freqs.reshape(-1, 1)], axis=1
            ).reshape(seq_len, -1)
        else:
            positions = mx.concatenate([freqs, freqs], axis=-1)

        pos_sin = mx.sin(positions)[None, None, :, :]
        pos_cos = mx.cos(positions)[None, None, :, :]
        return pos_sin, pos_cos

    def rotate_half(self, x):
        """Rotates half the hidden dims of the input."""
        x1 = x[..., : x.shape[-1] // 2]
        x2 = x[..., x.shape[-1] // 2 :]
        return mx.concatenate([-x2, x1], axis=-1)

    def rotate_every_two(self, x: mx.array) -> mx.array:
        B, nh, T, hs = x.shape
        x = x.reshape(B, nh, T, hs // 2, 2)
        x1, x2 = x[:, :, :, :, 0], x[:, :, :, :, 1]
        x = mx.stack([-x2, x1], axis=-1)
        return x.reshape(B, nh, T, hs)

    def apply_rotary_pos_emb(
        self, pos_sin: mx.array, pos_cos: mx.array, t: mx.array
    ) -> mx.array:
        if self.config.rope_impl == "interleave":
            return (t * pos_cos) + (self.rotate_every_two(t) * pos_sin)
        else:
            return (t * pos_cos) + (self.rotate_half(t) * pos_sin)

    def __call__(
        self, q: mx.array, k: mx.array, position_ids: Optional[mx.array] = None
    ) -> Tuple[mx.array, mx.array]:

        # Handle precision
        q_, k_ = (q, k)

        batch_size = q_.shape[0]
        query_len, key_len = q_.shape[-2], k_.shape[-2]

        if position_ids is not None:
            freqs_cis_len = self.config.max_position_embeddings
        else:
            freqs_cis_len = key_len

        pos_sin, pos_cos = self.get_rotary_embedding(freqs_cis_len)

        if position_ids is not None:
            assert (
                query_len == key_len
            ), "Query and key lengths must be equal when using position IDs."
            pos_sin = pos_sin[0, 0][position_ids].reshape(
                batch_size, 1, key_len, pos_sin.shape[-1]
            )
            pos_cos = pos_cos[0, 0][position_ids].reshape(
                batch_size, 1, key_len, pos_cos.shape[-1]
            )

        q_ = self.apply_rotary_pos_emb(
            pos_sin[:, :, key_len - query_len : key_len, :],
            pos_cos[:, :, key_len - query_len : key_len, :],
            q_,
        )
        k_ = self.apply_rotary_pos_emb(pos_sin, pos_cos, k_)

        return q_, k_
```

**Comment by awni** [comment]
Not sure.. can you share a command for how to run it and I can take a look?

**Comment by Blaizzy** [comment]
Here you go:
```
python -m mlx_vlm.generate --model allenai/Molmo-7B-D-0924 --max-token
s 50 --temp 0.1 --prompt "Hi, how are you?" --image "https://picsum.photos/id/237/536/354"
```

**Comment by Blaizzy** [comment]
By the way,

I disabled the vision path.

1. Commenting out the images 
Screenshot 2024-11-15 at 2 02 51 AM
2. Commenting out the model __call__ except:
Screenshot 2024-11-15 at 2 03 45 AM

**Comment by Blaizzy** [comment]
This model is not yet in HF transformers format so it's a bit tricky.

**Comment by Blaizzy** [comment]
@awni I figured it out!

This RoPe is a bit different.

The positions id has to be cache offset after prefill.

```python
 # Apply rotary embeddings
if cache is not None and cache.offset > 0:
    position_ids = mx.array([cache.offset])
else:
    position_ids = mx.arange(seq_len)

q, k = self.rotary_emb(q, k, position_ids)
     
```

File: mlx_vlm/models/molmo/language.py [hunk 1]
```diff
@@ -0,0 +1,239 @@
+import inspect
+import math
```
Comment (Blaizzy) [inline mlx_vlm/models/molmo/language.py]:
```suggestion
```

**Review by Blaizzy — COMMENTED** [review]
LGTM
